diff --git a/ucm2/USB-Audio/Focusrite/Scarlett-2i.conf b/ucm2/USB-Audio/Focusrite/Scarlett-2i.conf
index 7492e574..fa032297 100644
--- a/ucm2/USB-Audio/Focusrite/Scarlett-2i.conf
+++ b/ucm2/USB-Audio/Focusrite/Scarlett-2i.conf
@@ -55,6 +55,15 @@ If.gen3 {
 	True.Define.Generation "3rd"
 }
 
+If.gen4 {
+	Condition {
+		Type RegexMatch
+		String "${CardComponents}"
+		Regex "USB1235:821[89]"
+	}
+	True.Define.Generation "4th"
+}
+
 Comment "Focusrite Scarlett ${var:Model} ${var:Generation} Gen"
 
 SectionUseCase."HiFi" {
diff --git a/ucm2/USB-Audio/USB-Audio.conf b/ucm2/USB-Audio/USB-Audio.conf
index fa147d9c..7c8e5627 100644
--- a/ucm2/USB-Audio/USB-Audio.conf
+++ b/ucm2/USB-Audio/USB-Audio.conf
@@ -267,7 +267,9 @@ If.focusrite-scarlett-2i {
 		# 8205 Solo 2nd Gen
 		# 8210 2i2 3rd Gen
 		# 8211 Solo 3rd Gen
-		Regex "USB1235:8(0(0[6a]|1c)|2(0[025]|1[01]))"
+		# 8218 Solo 4th Gen
+		# 8219 2i2 4th Gen
+		Regex "USB1235:8(0(0[6a]|1c)|2(0[025]|1[0189]))"
 	}
 	True.Define {
 		ProfileName "Focusrite/Scarlett-2i"
